src/ssl.c

```diff
@@ -30936,6 +30985,230 @@ int wolfSSL_get_state(const WOLFSSL* ssl)
+        WOLFSSL_MSG("one or more inputs were NULL");
+        return BAD_FUNC_ARG;
+    }
+
+    if((bio = BIO_new(BIO_s_mem())) == NULL){
+        WOLFSSL_MSG("Set bio Failed");
+        return WOLFSSL_FAILURE;
+    }
+
+    pkey = wolfSSL_PKEY_new_ex(bio->heap);
+    pkey->type   = EVP_PKEY_RSA;
+    pkey->rsa    = rsa;
+    pkey->ownRsa = 0;
+
+    /* convert RSA internal to der */
+    if((derSize = wolfSSL_i2d_RSAPrivateKey(rsa, &derBuf)) <= 0) {
+        WOLFSSL_MSG("wolfSSL_i2d_RSAPrivateKey failed");
+        XFREE((unsigned char*)derBuf, NULL, DYNAMIC_TYPE_TMP_BUFFER);
+        wolfSSL_EVP_PKEY_free(pkey);
+        wolfSSL_BIO_free(bio);
+        return WOLFSSL_FAILURE;
+    }
+
+    pkey->pkey.ptr = (char*)XMALLOC(derSize, bio->heap, DYNAMIC_TYPE_TMP_BUFFER);
+    if (pkey->pkey.ptr == NULL) {
+        WOLFSSL_MSG("pkey malloc failed");
+        XFREE((unsigned char*)derBuf, NULL, DYNAMIC_TYPE_TMP_BUFFER);
+        wolfSSL_EVP_PKEY_free(pkey);
+        wolfSSL_BIO_free(bio);
+        return WOLFSSL_FAILURE;
+    }
+
+    pkey->pkey_sz = derSize;
+    XMEMCPY(pkey->pkey.ptr, derBuf, derSize);
+
+    /*adds the private key rsa to ctx.*/
+    ret = wolfSSL_CTX_use_PrivateKey(ctx, pkey);
+
+    wolfSSL_BIO_free(bio);
+    wolfSSL_EVP_PKEY_free(pkey);
+    XFREE((unsigned char*)derBuf, NULL, DYNAMIC_TYPE_TMP_BUFFER);
+    return ret;
+}
+#endif /* NO_RSA && !HAVE_FAST_RSA */
+
+
+WOLFSSL_EVP_PKEY* wolfSSL_d2i_PrivateKey_bio(WOLFSSL_BIO* bio, WOLFSSL_EVP_PKEY** out)
+{
+    const unsigned char* mem;
+    const unsigned char* memptr;
+    int memSz;
+    WOLFSSL_EVP_PKEY* key = NULL;
+    int type;
+    int algoID = 0;
+
+    WOLFSSL_ENTER("wolfSSL_d2i_PrivateKey_bio()");
+
+    if (bio == NULL) {
+        return NULL;
+    }
+    (void)out;
+
+    memSz = wolfSSL_BIO_pending(bio);
+    if (memSz <= 0) {
+        WOLFSSL_MSG("wolfSSL_BIO_pending failure");
+        return NULL;
+    }
+    mem = (unsigned char*)XMALLOC(memSz, bio->heap, DYNAMIC_TYPE_TMP_BUFFER);
+    if (mem == NULL) {
+        WOLFSSL_MSG("Malloc failure");
+        return NULL;
+    }
+    memptr = mem;
+
+    if (wolfSSL_BIO_read(bio, (unsigned char*)mem, memSz) == memSz) {
+        /* Determines key type */
+        if ((wc_GetKeyOID((unsigned char*)mem, memSz, NULL, NULL, &algoID, NULL)) < 0) {
+            XFREE((unsigned char*)memptr, bio->heap, DYNAMIC_TYPE_TMP_BUFFER);
+            return NULL;
+        }
+
+        if (algoID == ECDSAk) {
+            type = EVP_PKEY_EC;
+        } else if (algoID == RSAk) {
+            type = EVP_PKEY_RSA;
+        } else {
+            WOLFSSL_MSG("Error determining correct key type");
+            XFREE((unsigned char*)memptr, bio->heap, DYNAMIC_TYPE_TMP_BUFFER);
+            return NULL;
+        }
+
+        key = wolfSSL_d2i_PrivateKey(type, &key, &mem, memSz);
+        if (out != NULL && key != NULL) {
+            *out = key;
+        }
+    }
+    XFREE((unsigned char*)memptr, bio->heap, DYNAMIC_TYPE_TMP_BUFFER);
```

Comment:
at the end of the function we need to take into account data at  the end of the WOLFSSL_BIO buffer that may not be related to DER

Reply:
I'm little stuck on this one, any ideas?